### PR TITLE
2021-assignment-sklearn: Add new test cases

### PR DESCRIPTION
**Pull request by h2o64**
This PR adds many news tests

- Tests with an even number of months in the dataset
- Tests with shuffled input data
- Tests with resampled data (holes in the dataset)
- Tests to actually take `n_neighbors` into considerations

They leverage `pytest.mark.parametrize` to reuse the old tests as much as possible.

### CODE REVIEW

File: test_sklearn_questions.py [hunk 1]
```diff
@@ -111,3 +111,118 @@ def test_time_split_on_column():
...
+        assert X_train['date'].dt.year.nunique() == 1
+        assert X_test['date'].dt.year.nunique() == 1
+        if last_time is not None:
+            assert X_test['date'].min() > last_time
+        last_time = X_test['date'].max()
+        n_splits += 1
+
+    assert 'idx' not in X.columns
+
+    assert n_splits == cv.get_n_splits(X, y)
```
Comment (tomMoral) [inline test_sklearn_questions.py:228]: Missing enline.

**Review by tomMoral — COMMENTED** [review]
Could you try to add these tests without duplicating the code using [`pytest.mark.parametrize`](https://docs.pytest.org/en/6.2.x/parametrize.html)?

For instance add a `shuffle` parameter that is a boolean in the `test_time_split` and only shuffle `X` if it is true? Same you could add a `end` parameter for teh other two tests.

As for the 3rd test, I was thinking on finding some extra edge case.

**Comment by tomMoral** [comment]
For the 3rd test case, maybe you can find a edge case when the data is not continuous.
What if we are missing one month?

File: test_sklearn_questions.py [hunk 2]
```diff
@@ -35,15 +37,23 @@ def test_one_nearest_neighbor_check_estimator(k):
...
 
-def test_time_split():
+@pytest.mark.parametrize("end_date_and_splits",
+                         [('2021-01-31', 12), ('2020-12-31', 11)])
+@pytest.mark.parametrize("shuffle_data", [True, False])
+def test_time_split(end_date_and_splits, shuffle_data):
 
-    date = pd.date_range(start='2020-01-01', end='2021-01-31', freq='M')
+    date = pd.date_range(start='2020-01-01',
+                         end=end_date_and_splits[0], freq='M')
```
Comment (tomMoral) [inline test_sklearn_questions.py:46]: Make it a bit harder, to make sure we have multiple sample per split.
```suggestion
                         end=end_date_and_splits[0], freq='D')
```

File: test_sklearn_questions.py [hunk 3]
```diff
@@ -35,15 +37,23 @@ def test_one_nearest_neighbor_check_estimator(k):
     check_estimator(KNearestNeighbors(n_neighbors=k))
 
 
-def test_time_split():
+@pytest.mark.parametrize("end_date_and_splits",
+                         [('2021-01-31', 12), ('2020-12-31', 11)])
+@pytest.mark.parametrize("shuffle_data", [True, False])
+def test_time_split(end_date_and_splits, shuffle_data):
```
[the remark this answers is not in the document]
Comment (h2o64) [reply test_sklearn_questions.py:43]: Great suggestion ! I didn't know about this behavior

File: test_sklearn_questions.py [hunk 4]
```diff
@@ -35,15 +37,23 @@ def test_one_nearest_neighbor_check_estimator(k):
...
 
-def test_time_split():
+@pytest.mark.parametrize("end_date, expected_splits",
+                         [('2021-01-31', 12), ('2020-12-31', 11)])
+@pytest.mark.parametrize("shuffle_data", [True, False])
+def test_time_split(end_date, expected_splits, shuffle_data):
 
-    date = pd.date_range(start='2020-01-01', end='2021-01-31', freq='M')
+    date = pd.date_range(start='2020-01-01',
+                         end=end_date, freq='D')
```
Comment (tomMoral) [inline test_sklearn_questions.py:46]:
```suggestion
    date = pd.date_range(start='2020-01-01', end=end_date, freq='D')
```

**Review by tomMoral — COMMENTED** [review]
That is nice! I think the good thing would be to raise an error if there is a hole in the data.

File: test_sklearn_questions.py [hunk 5]
```diff
@@ -78,17 +88,29 @@ def test_time_split():
...
     )
 
+    if shuffle_data:
+        X, y = shuffle(X, y, random_state=0)
+
+    if sample_perc < 1.0:
+        sample_index = np.random.randint(X.shape[0],
+                                         size=int(sample_perc * X.shape[0]))
+        X = X.iloc[sample_index]
+        y = y.iloc[sample_index]
```
Comment (tomMoral) [inline test_sklearn_questions.py:112]:
```suggestion
```

File: test_sklearn_questions.py [hunk 6]
```diff
@@ -78,17 +88,29 @@ def test_time_split():
         next(cv.split(X, y))
 
 
-def test_time_split_on_column():
+@pytest.mark.parametrize("end_date", ['2021-01-31', '2020-12-31'])
+@pytest.mark.parametrize("shuffle_data", [True, False])
+@pytest.mark.parametrize("sample_perc", [0.2, 0.5, 0.8, 1.0])
+def test_time_split_on_column(end_date, shuffle_data, sample_perc):
```
Comment (tomMoral) [inline test_sklearn_questions.py:94]:
```suggestion
def test_time_split_on_column(end_date, shuffle_data):
```

Comment (tomMoral) [reply test_sklearn_questions.py:94]: Let's skip the 3rd test for now.

File: test_sklearn_questions.py [hunk 7]
```diff
@@ -78,17 +87,23 @@ def test_time_split():
...
     n_samples = len(date)
     X = pd.DataFrame({'val': range(n_samples), 'date': date})
     y = pd.DataFrame(
         np.array([i % 2 for i in range(n_samples)])
     )
 
+    if shuffle_data:
+        X, y = shuffle(X, y, random_state=0)
+
+
```
Comment (tomMoral) [inline test_sklearn_questions.py:106]:
```suggestion
```

**Review by tomMoral — APPROVED** [review]
LGTM! thx @h2o64
